### ISSUE TITLE
Incorrect timeout raised in response to cosocket timeout

### ISSUE THREAD

**Issue by Araneidae** [dls-controls/cothread#14]
The cothread `cosocket` module endeavours to emulate the underlying `socket` module as closely as possible.  Unfortunately our treatment of timeout differs: `socket` methods raise an instance of `socket.timeout` whereas cothread raises
```
socket.error(errno.ETIMEDOUT, 'Timeout waiting for socket')
```
Unfortunately fixing this is a breaking change...

**Comment by Araneidae** [comment]
@thomascobb , @willrogers , I'd appreciate your thoughts on this.

**Comment by thomascobb** [comment]
Fix it and make a 3.0 release?

**Comment by Araneidae** [comment]
I'm not keen for a major version step for such a trivial update.  I'll explain the precise details and consequences here.

First of all, the error is that `cosocket.py` raises `socket.error(errno.ETIMEDOUT, ...)` instead of `socket.timeout('timed out')`.  The exception `socket.error` is a superclass of `socket.timeout`.  The impact of this is shown by the following code:
```python
try:
    action_raising_timeout()
except socket.timeout:
    handle_timeout_1()
except socket.error as e:
    if e.errno == errno.ETIMEDOUT:
        handle_timeout_2()
    else:
        handle_socket_error()
```
With the current version of cothread `handle_timeout_2` will be called, or lacking the `errno` test, `handle_socket_error`; after fixing this discrepancy `handle_timeout_1` will be called, or if missing, `handle_socket_error` again.

The main problem is going to be with code that currently does the `if e.errno == errno.ETIMEDOUT` dance without catching `socket.timeout`: this code will see timeouts elevated to general socket errors.

**Comment by AlexanderWells-diamond** [comment]
This was fixed in commit 7e3557be205a3ee0661fa3979d7f1c0e5517d9dc
